code/extensions/che-remote/src/extension.ts

```diff
@@ -202,6 +205,21 @@ async function updateDevfile(cheApi: any): Promise<boolean> {
     return false;
   }
 
+  // if a new Devfile does not contain projects, copy them from flattened Devfile
+  try {
+    let projects: V1alpha2DevWorkspaceSpecTemplateProjects[] | undefined = devfileContext.devWorkspace.spec!.template!.projects;
+    if (!projects || projects.length === 0) {
```

Comment:
Something to consider:

When using the [reproducer repo](https://github.com/Nykredit/openshift-test/tree/single_root), if you add  a new project to your devfile, say:
```YAML
projects:
  - name: web-nodejs-sample
    git:
      remotes:
        origin: "https://github.com/che-samples/web-nodejs-sample.git"
```
And then restart from the local devfile, you'll have a single-root workspace that only contains the newly added `web-nodejs-sample` project. The existing project that we used to restart the workspace from local devfile is no longer shown in VSCode, since its project (that used to be present in the devworkspace) wasn't added to the newly generated devworkspace's projects.

So I suggest:
- _Always_ adding the devworkspace's existing projects `(flattenedDevfile.projects)`  to the  newly generated devworkspace's projects (`devfileContext.devWorkspace.spec!.template!.projects`) instead of only adding it _if_ the new devfile does not contain projects.
- Add the devworkspace's existing project(s) to be first in list of projects, since this is the project we had used to restart the workspace from local devfile (it should remain first after we restart)
- Don't add projects with the same name (e.g. if the same project name is used in the devworkspace and local devfile) because this will render the devworkspace invalid.

For reference, the approach I'm suggesting is implemented in DevWorkspace Operator [here](https://github.com/devfile/devworkspace-operator/pull/1193/files#diff-879eb83b19c3051f7f7a0a65d079b4bc52a7891b512757420123a113c03e00c7R91-R123).

Reply:
I would give the user all the freedom.
If he added a project to his devfile, he would get only one project in the tree.
Adding a project with the same name should be possible and handled when cloning the project.

In this PR I handled only a case when the user created a workspace from minimal devfile and he wants only to change the devfile a bit and apply the changes.

I would not fix all the cases in this PR. Let's move step by step.
You can check some discussion here https://github.com/eclipse-che/che/issues/22722

Reply:
Sorry for slow response but yes this makes sense. If changes are needed, we can discuss them in https://github.com/eclipse-che/che/issues/23036  :+1: